src/program/latex/mod.rs

```diff
@@ -0,0 +1,1291 @@
+    /// the wire. If the gate name is a composite gate, the gate name is
+    /// decomposed into canonical form before being set in the circuit. For
+    /// example, CNOT is a composite gate that can be decomposed into the
+    /// equivalent canonical form, CONTROLLED X.
+    ///
+    /// # Arguments
+    /// `self` - exposes all attributes in the diagram
+    /// `gate` - the Gate of the Instruction from `to_latex`.
+    fn parse_gate(&mut self, gate: &Gate) -> Result<(), LatexGenError> {
+        // preserve qubit order in instruction
+        self.relationships.insert(self.column, gate.qubits.clone());
+
+        // set modifiers from gate instruction
+        for qubit in &gate.qubits {
+            if let Qubit::Fixed(qubit) = qubit {
+                if let Some(wire) = self.circuit.get_mut(qubit) {
+                    // set modifiers at this column for all qubits
+                    Self::set_modifiers(wire, &self.column, &gate.modifiers)?;
+
+                    // set parameters at this column for all qubits
+                    for expression in &gate.parameters {
+                        wire.set_param(
+                            expression,
+                            self.column,
+                            self.settings.texify_numerical_constants,
+                        );
+                    }
+                }
+            }
+        }
+
+        // parse the gate to a canonical gate if supported
+        let canonical_gate = match CanonicalGate::get_canonical(&gate.name) {
+            CanonicalGate::Cnot(inst) => Some(inst),
+            CanonicalGate::Ccnot(inst) => Some(inst),
+            CanonicalGate::Cphase(inst) => Some(inst),
+            CanonicalGate::Cz(inst) => Some(inst),
+            CanonicalGate::None => None,
+        };
+
+        // add the qubits to the canonical gate to form an instruction
+        let instruction = if let Some(mut canonical_gate) = canonical_gate {
+            for qubit in &gate.qubits {
+                if let Qubit::Fixed(qubit) = qubit {
+                    canonical_gate.push(' ');
+                    canonical_gate.push_str(&qubit.to_string());
+                }
+            }
+            Some(canonical_gate)
+        } else {
+            None
+        };
+
+        // get gate from new program of canonical instruction
+        if let Some(instruction) = instruction {
+            let instructions = Program::from_str(&instruction)
+                .expect("should return program {instruction}")
+                .to_instructions(false);
+
+            for instruction in instructions {
+                if let Instruction::Gate(gate) = instruction {
+                    // call until all composite gates are in canonical form
+                    self.parse_gate(&gate)?;
+                }
+            }
+        // gate is in canonical form, build wire
+        } else {
+            // set gates
```

Comment:
I'm considering moving this block into a separate function called `set_gate`. However, I want to keep one of the arguments as `&mut self` but since this function, `parse_gate`, already has `&mut self`, I'm running into this compilation error: `cannot borrow `*self` as mutable because it is also borrowed as immutable`. My only way around this to set the function signature using mutable references from the fields I want to update in `Diagram` but that's not very elegant which is why I haven't moved this block yet.

Reply:
Moved it to a method in `Wire` called `set_gate`. [Refactor gate setting from parse_gate into separate method set_gate.](https://github.com/rigetti/quil-rs/pull/145/commits/029f05b6c251b8e1195fa5e63212d6f563c19a2d)